### PR TITLE
Postprocess file

### PR DESCRIPTION
**Pull request by dpgrote**
Rewrites and cleans up the plot_parallel.py script. This includes my suggested changes, as well as several other changes.

### CODE REVIEW

File: Tools/plot_parallel.py [hunk 1]
```diff
@@ -20,37 +24,39 @@
...
                     help='Only particles with -slicewidth/2<y<slicewidth/2 are plotted')
-parser.add_argument('--parallel', dest='parallel', action='store_true', default=False,
+parser.add_argument('--parallel', action='store_true', default=False,
                     help='whether or not to do the analysis in parallel (e.g., 1 plotfile per MPI rank)')
 parser.add_argument('--species', dest='pslist', nargs='+', type=str, default=None,
                     help='Species to be plotted, e.g., " --species beam plasma_e ". By default, all species in the simulation are shown')
-parser.add_argument('--plot_Ey_max_evolution', dest='plot_Ey_max_evolution', action='store_true', default=False,
-                    help='Whether to plot evolution of max(Ey) to illustrate how to plot one quantity across all plotfiles.')
+parser.add_argument('--plot_max_evolution', type=str, default=None,
+                    help='Quantity to plot the max of across all data files')
```
Comment (MaxThevenet) [inline Tools/plot_parallel.py]: Right now I am using the script to plot the evolution of particle quantities... Do you think we should rename this option and the functions involved with generic names, e.g., `--plot_evolution` and replace function `get_field_max` by `evolution_get_quantity` etc.?
If so, that could be included in this PR or we can make another PR for that (or just create an issue).

File: Tools/plot_parallel.py [hunk 2]
```diff
@@ -100,25 +106,27 @@ def plot_snapshot(filename):
...
+        if vmax is not None:
+            sl.set_zlim(-vmax, vmax)
 
     # Plot particle quantities
     for ispecies, pspecies in enumerate(pslist):
         if pspecies in [x[0] for x in ds.field_list]:
             if plotlib == 'matplotlib':
                 # Read particle quantities from yt dataset
-                xp = ad[pspecies, 'particle_position_x'].v
+                xp = all_data_level_0[pspecies, 'particle_position_x'].v
```
Comment (MaxThevenet) [inline Tools/plot_parallel.py]: I understand that there was a bug here, but I am not sure this new version would work in the presence of MR. Could we revert from `all_data_level_0` to `ad` for particle quantities, and add `ad = ds.all_data()` right after `if plotlib == 'matplotlib':` (I accidentally deleted it)?
Thanks for catching it!

File: Tools/plot_parallel.py [hunk 3]
```diff
@@ -158,95 +163,64 @@ def get_field_max( filename, a_field ):
...
-    print('Serial analysis')
-    print('Number of plotfiles: %s' %nfiles)
-    pslist = get_species(file_list);
+    try:
+        from mpi4py import MPI
+        comm_world = MPI.COMM_WORLD
+        rank = comm_world.Get_rank()
+        size = comm_world.Get_size()
+    except ImportError:
+        pass
```
Comment (MaxThevenet) [inline Tools/plot_parallel.py]: To my understanding, if the user does not have `mpi4py` available, and uses `python plot_parallel.py`, this would run serial successfully, right?

File: Tools/plot_parallel.py [hunk 4]
```diff
@@ -135,15 +143,12 @@ def plot_snapshot(filename):
     if plotlib == 'matplotlib':
         plt.xlabel('z (m)')
         plt.ylabel('x (m)')
-        plt.title(args.field + ' at iteration ' + str(iteration) +
-                  ', time = ' + str(ds.current_time))
-        plt.savefig(args.path + '/plt_' + args.field + '_' + plotlib + '_' +
-                    str(iteration).zfill(5) + '.png', bbox_inches='tight', dpi=300)
+        plt.title('%s at iteration %d, time = %e s'%(args.field, iteration, ds.current_time))
+        plt.savefig('plt_%s_%s_%05d.png'%(args.field, plotlib, iteration), bbox_inches='tight', dpi=300)
```
Comment (MaxThevenet) [inline Tools/plot_parallel.py]: Did you remove `args.path + ` on purpose? As far as I am concerned, I like when the results are saved close to the data. Alternatively, do you think it would be worth adding an option
```py
parser.add_argument('--path_img', default='.',
	                    help='path of directory where images are saved.')
```
the default can be `.` (new version) or `drags/plotfiles` (old version).

File: Tools/plot_parallel.py [hunk 5]
```diff
@@ -135,15 +143,12 @@ def plot_snapshot(filename):
...
         plt.close()
     if plotlib == 'yt':
         sl.annotate_grids()
-        sl.save(args.path + '/plt_' + args.field + '_' + plotlib + '_' +
-                str(iteration).zfill(5) + '.png')
+        sl.save('plt_%s_%s_%05d.png'%(args.field, plotlib, iteration))
```
Comment (MaxThevenet) [inline Tools/plot_parallel.py]: Same comment as above concerning images directory.

File: Tools/plot_parallel.py [hunk 6]
```diff
@@ -158,95 +163,64 @@ def get_field_max( filename, a_field ):
...
-def plot_field_max():
+def plot_field_max(zwin_arr, maxF_arr):
     plt.figure()
     plt.plot(zwin_arr, maxF_arr)
     plt.xlabel('z (m)')
-    plt.ylabel('Field (S.I.)')
+    plt.ylabel('%s (S.I.)'%plot_max_evolution)
     plt.title('Field max evolution')
-    plt.savefig('max_field_evolution.pdf', bbox_inches='tight')
+    plt.savefig('max_%s_evolution.pdf'%plot_max_evolution, bbox_inches='tight')
```
Comment (MaxThevenet) [inline Tools/plot_parallel.py]: Same comment as above concerning images directory.

File: Tools/plot_parallel.py [hunk 7]
```diff
@@ -158,95 +163,64 @@ def get_field_max( filename, a_field ):
...
+    plot_snapshot( filename )
+    if plot_max_evolution is not None:
+        zwin[count], maxF[count] = get_field_max( filename, plot_max_evolution )
+
+if plot_max_evolution is not None:
+    if size > 1:
+        global_zwin = np.empty_like(zwin)
+        global_maxF = np.empty_like(maxF)
+        comm_world.Reduce(zwin, global_zwin, op=MPI.MAX)
+        comm_world.Reduce(maxF, global_maxF, op=MPI.MAX)
```
Comment (MaxThevenet) [inline Tools/plot_parallel.py]: Ok my bad, this does give the expected result (sorry I don't get what the Reduce function does).

File: Tools/plot_parallel.py (hunk 1 above)
Comment (dpgrote) [reply Tools/plot_parallel.py]: Yes, the names can be changed to be generic. I'll make the change.

File: Tools/plot_parallel.py (hunk 2 above)
Comment (dpgrote) [reply Tools/plot_parallel.py]: Yes, this can be changed to use ds.all_data()

File: Tools/plot_parallel.py (hunk 3 above)
Comment (dpgrote) [reply Tools/plot_parallel.py]: Correct.

I think it would be better to change this so that it is parallel by default and replace the parallel flag with a serial flag.

File: Tools/plot_parallel.py (hunk 4 above)
Comment (dpgrote) [reply Tools/plot_parallel.py]: I can add the args.path prefix back in. My preference is that the images appear in the directory where this is used, but putting them with the plotfiles is fine. I can add the optional path, though I would call it "image_dir".

File: Tools/plot_parallel.py (hunk 7 above)
Comment (dpgrote) [reply Tools/plot_parallel.py]: For each element in the array passed in, reduce finds the maximum value of the element over the processors.

File: Tools/plot_parallel.py (hunk 3 above)
Comment (MaxThevenet) [reply Tools/plot_parallel.py]: I agree

File: Tools/plot_parallel.py (hunk 4 above)
Comment (MaxThevenet) [reply Tools/plot_parallel.py]: Ok, perfect.

File: Tools/plot_parallel.py (hunk 7 above)
Comment (MaxThevenet) [reply Tools/plot_parallel.py]: Oh I get it now, it is actually an element-wise `Reduce` operation, and for a given index all ranks have `np.finfo(float).min` except the rank that operated on the index (which has a >0 value).

Then this would break if the user is interested in a quantity that can be <0. If a user needs the average transverse position of the beam (which I am actually doing now), all negative values will be replaced by zeros. Could we add a safeguard for this?

Comment (dpgrote) [reply Tools/plot_parallel.py]: np.finfo(float).min is the lowest, most negative, float, typically -1.7976931348623157e+308.
